sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -3146,82 +3146,96 @@ atomic_update(T *p, Toffset offset, simd<T, N> src0, simd<T, N> src1,
 
 template <native::lsc::atomic_op Op, typename T, int N, typename Toffset,
           typename AccessorTy>
-__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset> &&
-                                 __ESIMD_DNS::get_num_args<Op>() == 0 &&
-                                 !std::is_pointer_v<AccessorTy>,
-                             simd<T, N>>
-atomic_update(AccessorTy acc, simd<Toffset, N> offset, simd_mask<N> mask) {
+__ESIMD_API
+    std::enable_if_t<std::is_integral_v<Toffset> &&
+                         __ESIMD_DNS::get_num_args<Op>() == 0 &&
+                         (__ESIMD_DNS::is_rw_device_accessor_v<AccessorTy> ||
+                          __ESIMD_DNS::is_rw_local_accessor_v<AccessorTy>),
```

Comment:
Nit: We could make a is_rw_device_or_local_accessor_v function if we expect to use this logic in a lot of places